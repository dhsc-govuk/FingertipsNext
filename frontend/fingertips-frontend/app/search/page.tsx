import { SearchForm } from '@/components/forms/SearchForm';

export default async function Page(
  props: Readonly<{
    searchParams?: Promise<{
      indicator?: string;
    }>;
  }>
) {
  const searchParams = await props.searchParams;
  const indicator = searchParams?.indicator ?? '';

  return <SearchForm indicator={indicator} />;
}
