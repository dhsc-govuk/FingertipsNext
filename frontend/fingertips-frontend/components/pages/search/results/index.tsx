'use client';

import { BackLink, H1, Paragraph } from 'govuk-react';

export function SearchResults({ indicator }: Readonly<{ indicator: string }>) {
  return (
    <>
      <BackLink href={`/search?indicator=${indicator}`} />
      <H1>Search results</H1>
      <Paragraph>{`You searched for indicator "**${indicator}**"`}</Paragraph>
    </>
  );
}
