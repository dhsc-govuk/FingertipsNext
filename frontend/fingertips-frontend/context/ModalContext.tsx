import {
  createContext,
  FC,
  ReactNode,
  useContext,
  useMemo,
  useState,
} from 'react';
import { ModalPane } from '@/components/molecules/ModalPane/ModalPane';

export interface Modal {
  content?: ReactNode;
  title?: string;
}

export const ModalContext = createContext<{
  modal: Modal;
  setModal: (newPreview: Modal) => void;
}>({ modal: {}, setModal: () => undefined });

export const useModal = () => {
  const { modal, setModal } = useContext(ModalContext);
  const { content, title } = modal;
  return { modal, setModal, content, title };
};

interface ModalProviderProps {
  children: ReactNode;
}

export const ModalProvider: FC<ModalProviderProps> = ({ children }) => {
  const [modal, setModal] = useState<Modal>({});
  const providerValue = useMemo(
    () => ({
      modal,
      setModal,
    }),
    [modal]
  );

  return (
    <ModalContext.Provider value={providerValue}>
      {children}
      <ModalPane />
    </ModalContext.Provider>
  );
};
