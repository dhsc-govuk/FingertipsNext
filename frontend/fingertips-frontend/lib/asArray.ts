export const asArray = (value?: string | string[]): string[] => {
  if (value) {
    return Array.isArray(value) ? value : [value];
  }
  return [];
};
