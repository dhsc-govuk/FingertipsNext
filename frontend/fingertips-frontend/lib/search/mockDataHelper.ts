import { IndicatorDocument } from './searchTypes';

export const generateIndicatorDocument = (id: string): IndicatorDocument => ({
  indicatorID: id,
  indicatorName: `indicator name for id ${id}`,
  indicatorDefinition: `Some definition for id ${id}`,
  dataSource: `Some data source for id ${id}`,
  earliestDataPeriod: '2022',
  latestDataPeriod: '2023',
  lastUpdatedDate: new Date(),
  associatedAreaCodes: [],
  unitLabel: 'some unit label',
  hasInequalities: true,
  usedInPoc: true,
});
