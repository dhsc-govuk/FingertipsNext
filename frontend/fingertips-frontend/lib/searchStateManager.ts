import { asArray } from './pageHelpers';

export enum SearchParams {
  SearchedIndicator = 'si',
  IndicatorsSelected = 'is',
  AreasSelected = 'as',
  AreaTypeSelected = 'ats',
  GroupTypeSelected = 'gts',
  ConfidenceIntervalSelected = 'cis',
}

export type SearchParamKeys = `${SearchParams}`;

const multiValueParams = [
  SearchParams.IndicatorsSelected as string,
  SearchParams.AreasSelected as string,
  SearchParams.ConfidenceIntervalSelected as string,
];

export type SearchStateParams = {
  [SearchParams.SearchedIndicator]?: string;
  [SearchParams.IndicatorsSelected]?: string[];
  [SearchParams.AreasSelected]?: string[];
  [SearchParams.AreaTypeSelected]?: string;
  [SearchParams.GroupTypeSelected]?: string;
  [SearchParams.ConfidenceIntervalSelected]?: string[];
};

const isMultiValueTypeParam = (searchParamKey: SearchParamKeys) =>
  multiValueParams.includes(searchParamKey);

export class SearchStateManager {
  private searchState: SearchStateParams;
  private searchStateParams: URLSearchParams;

  constructor(searchState: SearchStateParams = {}) {
    this.searchState = {};
    this.searchStateParams = new URLSearchParams();

    this.searchState = searchState;
  }

  private constructPath(path: string) {
    if (this.searchStateParams.size === 0) {
      return path;
    }
    return `${path}?${this.searchStateParams.toString()}`;
  }

  private addParamToPath(searchParamKey: SearchParamKeys) {
    const paramsToAdd = asArray(this.searchState[searchParamKey]);

    if (paramsToAdd.length > 0) {
      paramsToAdd.forEach((paramToAdd) => {
        this.searchStateParams.append(searchParamKey, paramToAdd);
      });
    }
  }

  public addParamValueToState(
    searchParamKey: SearchParamKeys,
    paramToAdd: string
  ) {
    if (isMultiValueTypeParam(searchParamKey)) {
      const currentParamState = this.searchState[searchParamKey];

      const currentParamStateAsArray = asArray(currentParamState);
      const doesParamExist = currentParamStateAsArray.find(
        (currentParam) => currentParam === paramToAdd
      );

      if (!doesParamExist) {
        const newState: SearchStateParams = {
          ...this.searchState,
          [searchParamKey]: [...currentParamStateAsArray, paramToAdd],
        };

        this.searchState = newState;
      }
    } else {
      const newState: SearchStateParams = {
        ...this.searchState,
        [searchParamKey]: paramToAdd,
      };

      this.searchState = newState;
    }
  }

  public removeParamValueFromState(
    searchParamKey: SearchParamKeys,
    paramToRemove?: string
  ) {
    if (isMultiValueTypeParam(searchParamKey)) {
      const currentParamState = asArray(this.searchState[searchParamKey]);

      const newParamState = currentParamState?.filter((paramValue) => {
        return paramValue !== paramToRemove;
      });

      const newState: SearchStateParams = {
        ...this.searchState,
        [searchParamKey]: newParamState,
      };

      this.searchState = newState;
    } else {
      const newState: SearchStateParams = {
        ...this.searchState,
        [searchParamKey]: undefined,
      };

      this.searchState = newState;
    }
  }

  public removeAllParamFromState(searchParamKey: SearchParamKeys) {
    this.searchState[searchParamKey] = undefined;
  }

  public static setStateFromParams(params: URLSearchParams) {
    const newState = Object.values(SearchParams).reduce<SearchStateParams>(
      (state, searchParamKey) => {
        if (isMultiValueTypeParam(searchParamKey)) {
          const paramValues = params.getAll(searchParamKey);

          if (paramValues.length > 0) {
            const newState: SearchStateParams = {
              ...state,
              [searchParamKey]: paramValues,
            };

            return newState;
          }
        } else {
          const paramValue = params.get(searchParamKey);

          if (paramValue) {
            const newState: SearchStateParams = {
              ...state,
              [searchParamKey]: paramValue,
            };

            return newState;
          }
        }
        return state;
      },
      {}
    );

    const searchStateManager = new SearchStateManager(newState);
    return searchStateManager;
  }

  public static setStateFromSearchStateParams(params?: SearchStateParams) {
    if (params) {
      const newState = Object.values(SearchParams).reduce<SearchStateParams>(
        (state, searchParamKey) => {
          if (isMultiValueTypeParam(searchParamKey)) {
            const paramValues = asArray(params[searchParamKey]);

            if (paramValues.length > 0) {
              const newState: SearchStateParams = {
                ...state,
                [searchParamKey]: paramValues,
              };

              return newState;
            }
          } else {
            const paramValue = params[searchParamKey];

            if (paramValue) {
              const newState: SearchStateParams = {
                ...state,
                [searchParamKey]: paramValue,
              };

              return newState;
            }
          }
          return state;
        },
        {}
      );

      const searchStateManager = new SearchStateManager(newState);
      return searchStateManager;
    }
    return new SearchStateManager();
  }

  public getSearchState() {
    return this.searchState;
  }

  public generatePath(path: string) {
    this.searchStateParams = new URLSearchParams();

    Object.values(SearchParams).forEach((searchParamKey) => {
      this.addParamToPath(searchParamKey);
    });

    return this.constructPath(path);
  }
}
