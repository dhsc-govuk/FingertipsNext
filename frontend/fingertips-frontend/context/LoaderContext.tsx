'use client';

import React, {
  ReactNode,
  createContext,
  useContext,
  useMemo,
  useState,
  useEffect,
  Suspense,
  Dispatch,
  SetStateAction,
} from 'react';
import { usePathname } from 'next/navigation';
import { LoadingBox } from 'govuk-react';
import { useSearchStateParams } from '@/components/hooks/useSearchStateParams';
import { SearchStateParams } from '@/lib/searchStateManager';

export type LoaderContext = {
  getIsLoading: () => boolean;
  setIsLoading: (loading: boolean) => void;
};

type LoaderContextProvider = {
  children: ReactNode;
};

const LoaderContext = createContext<LoaderContext | undefined>(undefined);

export const LoaderProvider: React.FC<LoaderContextProvider> = ({
  children,
}) => {
  const pathname = usePathname();
  const [searchState, setSearchState] = useState<SearchStateParams>({});

  const [isLoading, setIsLoading] = useState<boolean>(true);

  const contextValue: LoaderContext = useMemo(
    () => ({
      getIsLoading: () => isLoading,
      setIsLoading,
    }),
    [isLoading, setIsLoading]
  );

  useEffect(() => {
    setIsLoading(false);
  }, [setIsLoading, pathname, searchState]);

  return (
    <LoaderContext.Provider value={contextValue}>
      <Suspense>
        <SuspendedUseSearchStateParams setSearchState={setSearchState} />
      </Suspense>
      <LoadingBox loading={isLoading} timeIn={200} timeOut={200}>
        {children}
      </LoadingBox>
    </LoaderContext.Provider>
  );
};

export const useLoadingState = () => {
  const context = useContext(LoaderContext);

  if (!context) {
    throw new Error('useLoadingState must be used within a LoaderProvider');
  }

  return context;
};

// Next has a very peculiar quirk where useSearchState cannot be called server
// side, which works fine during dev but build will fail.
// Here we're creating a component which will not render server side by
// suspending it until client mounts at which point it can set the loading
// context value, all this is more complicated because loader context is in a
// layout.
function SuspendedUseSearchStateParams({
  setSearchState,
}: Readonly<{ setSearchState: Dispatch<SetStateAction<SearchStateParams>> }>) {
  const searchParams = useSearchStateParams();
  useEffect(() => {
    setSearchState(searchParams);
  }, [searchParams, setSearchState]);

  return null;
}
