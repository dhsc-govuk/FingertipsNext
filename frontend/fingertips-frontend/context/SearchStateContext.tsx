'use client';

import React, { ReactNode, createContext, useContext, useState } from 'react';
import { SearchStateParams } from '@/lib/searchStateManager';

export type SearchStateContext = {
  getSearchState: () => SearchStateParams;
  setSearchState: (state: SearchStateParams) => void;
};

type SearchStateContextProvider = {
  children: ReactNode;
};

const SearchStateContext = createContext<SearchStateContext | undefined>(
  undefined
);

export const SearchStateProvider: React.FC<SearchStateContextProvider> = ({
  children,
}) => {
  const [searchState, setSearchState] = useState<SearchStateParams>({});

  const contextValue: SearchStateContext = {
    getSearchState: () => searchState,
    setSearchState,
  };

  return (
    <SearchStateContext.Provider value={contextValue}>
      {children}
    </SearchStateContext.Provider>
  );
};

export const useSearchState = () => {
  const context = useContext(SearchStateContext);

  if (!context) {
    throw new Error('useLoadingState must be used within a LoaderProvider');
  }

  return context;
};
