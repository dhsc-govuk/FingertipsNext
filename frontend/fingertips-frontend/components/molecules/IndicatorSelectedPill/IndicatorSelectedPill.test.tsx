import { render, screen } from '@testing-library/react';
import { IndicatorSelectedPill } from './index';
import { SearchParams, SearchStateParams } from '@/lib/searchStateManager';
import { generateIndicatorDocument } from '@/lib/search/mockDataHelper';
import { LoaderContext } from '@/context/LoaderContext';
import { SearchStateContext } from '@/context/SearchStateContext';

const mockIndicator = generateIndicatorDocument('1');

const mockSetIsLoading = vi.fn();
const mockLoaderContext: LoaderContext = {
  getIsLoading: vi.fn(),
  setIsLoading: mockSetIsLoading,
};

vi.mock('@/context/LoaderContext', () => {
  return {
    useLoadingState: () => mockLoaderContext,
  };
});

const mockGetSearchState = vi.fn().mockReturnValue({
  [SearchParams.AreasSelected]: ['A001', 'A002'],
  [SearchParams.IndicatorsSelected]: ['1'],
} satisfies SearchStateParams);
const mockSearchStateContext: SearchStateContext = {
  getSearchState: mockGetSearchState,
  setSearchState: vi.fn(),
};
vi.mock('@/context/SearchStateContext', () => {
  return {
    useSearchState: () => mockSearchStateContext,
  };
});

describe('IndicatorSelectedPill', () => {
  it('should render the IndicatorSelectedPill correctly', () => {
    render(<IndicatorSelectedPill indicator={mockIndicator} />);
    expect(screen.getByText(mockIndicator.indicatorName)).toBeInTheDocument();
    expect(screen.getByRole('link')).toBeInTheDocument();
  });

  it('should have the correct href with searchState as query param to the indicator meta data page', async () => {
    const expectedPath = [
      `/indicator/${mockIndicator.indicatorID}`,
      `?${SearchParams.IndicatorsSelected}=${mockIndicator.indicatorID}`,
      `&${SearchParams.AreasSelected}=A001&${SearchParams.AreasSelected}=A002`,
    ].join('');

    render(<IndicatorSelectedPill indicator={mockIndicator} />);

    expect(screen.getByRole('link')).toHaveAttribute('href', expectedPath);
  });

  it('should match snapshot', () => {
    const { asFragment } = render(
      <IndicatorSelectedPill indicator={mockIndicator} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
