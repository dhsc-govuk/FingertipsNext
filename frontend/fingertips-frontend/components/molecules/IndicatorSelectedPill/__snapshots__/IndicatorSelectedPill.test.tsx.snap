// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IndicatorSelectedPill > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gEvEer cmUCgj"
    data-testid="pill-container"
  >
    <div
      class="sc-eqUAAy eqmima"
      data-testid="filter-name"
    >
      indicator name for id 1
      <br />
      <a
        class="src__Link-sc-1loawqx-0 sc-dcJsrY jjZtWC kXAAIg"
        data-testid="view-background-info-link"
        href="/indicator/1?is=1&as=A001&as=A002"
      >
        View background information
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`IndicatorSelectedPill should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  background-color: white;
  border: 1px #D1D2D3 solid;
  border-radius: 5px;
  padding: 0.5em 0.3125em 0.5em 0.3125em;
  max-width: 100%;
  margin: 0.3125em 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  word-wrap: break-word;
  padding: 0em 1em 0em 0.75em;
  font-size: 16px;
  margin-bottom: 0;
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.c2 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.c2:focus {
  outline: 3px solid #ffdd00;
  outline-offset: 0;
  background-color: #ffdd00;
}

.c2:link {
  color: #1d70b8;
}

.c2:visited {
  color: #4c2c92;
}

.c2:hover {
  color: #2b8cc4;
}

.c2:active {
  color: #2b8cc4;
}

.c2:focus {
  color: #0b0c0c;
}

.c3 {
  font-size: 16px;
}

@media print {
  .c1 {
    font-family: sans-serif;
  }
}

@media print {
  .c2 {
    font-family: sans-serif;
  }
}

@media print {
  .c2[href^="/"]::after,
  .c2[href^="http://"]::after,
  .c2[href^="https://"]::after {
    content: " (" attr(href) ")";
    font-size: 90%;
    word-wrap: break-word;
  }
}

<div
    class="c0"
    data-testid="pill-container"
  >
    <div
      class="c1"
      data-testid="filter-name"
    >
      indicator name for id 1
      <br />
      <a
        class="c2 c3"
        data-testid="view-background-info-link"
        href="/indicator/1?is=1&as=A001&as=A002"
      >
        View background information
      </a>
    </div>
  </div>
</DocumentFragment>
`;
